### PR TITLE
Fix credential-by-value

### LINKED ISSUE #273: Are we doing credential-by-value wrong?
The places where ID_CRED_x is used in the plaintexts is described in RFC9528 as `ID_CRED_R / bstr / -24..23`.

My understanding is that fundamentally, ID_CRED_x is always a 1-element map, and that the bstr / -24..23 are just the shorthands of [Section 3.5.3.2](https://datatracker.ietf.org/doc/html/rfc9528#compact-kid) that expand to `{4: bstr}` (where -24..23 are also translated to bstr, and regular bstr are used as-is). The RFC also registers dedicated keys for that map for kccs and kcwt (we currently only use CCSs).

What we send and parse when doing credential-by-value is that we take the CRED_x value (which is a byte string), encode it as a CBOR byte string, and include it in the plain text. This was recently refactored into IdCred::write_to_message in #267 (while not altering Lakers' behavior).

I believe that this is an error, and that instead of using 482E2E2E2E2E2E2E2E ('........') we should be using A10E2E2E2E2E2E2E2E2E ({/kccs/ 14: '........'}). The RFC9529 traces do not contain an instance of credential-by-value that I could find. Has sending a credential by value ever been plugtested?

### PR DESCRIPTION
**Pull request by geonnave**
Fixes #273.

### CODE REVIEW

File: examples/coap/src/bin/coapclient.rs [hunk 1]
```diff
@@ -53,7 +53,7 @@ fn client_handshake() -> Result<(), EDHOCError> {
 
     let mut msg_3 = Vec::from(c_r.as_cbor());
     let (mut initiator, message_3, prk_out) =
-        initiator.prepare_message_3(CredentialTransfer::ByReference, &None)?;
+        initiator.prepare_message_3(CredentialTransfer::ByValue, &None)?;
```
Comment (chrysn) [inline examples/coap/src/bin/coapclient.rs:56]: Do we want to change this here or is this just an artifact of debugging?

File: shared/src/lib.rs [hunk 2]
```diff
@@ -49,7 +49,7 @@ pub const ENCODED_VOUCHER_LEN: usize = 1 + MAC_LENGTH; // 1 byte for the length
 // maximum supported length of connection identifier for R
 pub const MAX_KDF_CONTEXT_LEN: usize = 150;
 pub const MAX_KDF_LABEL_LEN: usize = 15; // for "KEYSTREAM_2"
-pub const MAX_BUFFER_LEN: usize = 256;
+pub const MAX_BUFFER_LEN: usize = 256 + 32;
```
Comment (chrysn) [inline shared/src/lib.rs:52]: Is this intentional or just an artifact of debugging with credentials that happen to be a bit longer than needed? (A CRED_x that does not carry the full dash separated MAC address may be short enough)

File: shared/src/lib.rs [hunk 3]
```diff
@@ -937,6 +950,19 @@ mod cbor_decoder {
...
 
+        /// Begin decoding a map.
+        pub fn map(&mut self) -> Result<usize, CBORError> {
+            let b = self.read()?;
+            if CBOR_MAJOR_MAP != Self::type_of(b) {
+                Err(CBORError::DecodingError)
+            } else {
+                match Self::info_of(b) {
+                    31 => Err(CBORError::DecodingError), // no support for unknown size arrays
+                    n => Ok(self.as_usize(n)?),
```
Comment (chrysn) [inline shared/src/lib.rs:961]:
```suggestion
                    n if n < 24 => Ok(self.as_usize(n)?),
                    31 => Err(CBORError::DecodingError), // no support for long or indeterminate size
```

**Review by chrysn — COMMENTED** [review]
LGTM, didn't test yet. Comments inline.

**Comment by chrysn** [comment]
It appears that something is wrong with the commits. Could it be that your changes were committed --amend to the latest merge commit?

**Comment by chrysn** [comment]
I've taken the liberty to force-push a version that has an actual commit in there.

**Comment by chrysn** [comment]
I have now tested this and it works on self-interop. Let's see if we can even squeeze one more interop test into this hackathon.

File: shared/src/lib.rs (hunk 3 above)
Comment (geonnave) [reply shared/src/lib.rs]: Accepted.

File: examples/coap/src/bin/coapclient.rs (hunk 1 above)
Comment (geonnave) [reply examples/coap/src/bin/coapclient.rs]: Accepted.

File: shared/src/lib.rs (hunk 2 above)
Comment (geonnave) [reply shared/src/lib.rs]: Solved

**Comment by chrysn** [comment]
We're not there yet: from interop, we did `{/kccs/14: <<{...}>>}` while Marco (and the IANA registry) say it is `{/kccs/14: {...}}`

**Comment by geonnave** [comment]
Final part of the fix provided in PR #281.

**Pull request merged**